concepts/05 UI Components/DataGrid/10 Enhance Performance on Large Datasets/030 Rendering Optimization.md

```diff
@@ -1,7 +1,16 @@
+If you work with a large data set, apply the following recommendations to optimize rendering performance:
 
-Use virtual row rendering when [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is more than the default value.
-
-Virtual column rendering can be used when there are more than 20 columns outside the viewport.
+* **Render only visible rows and columns**             
+The DataGrid renders all rows and columns once the data is loaded. To optimize performance, set the [rowRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/rowRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#rowRenderingMode') and [columnRenderingMode](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/scrolling/columnRenderingMode.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/scrolling/#columnRenderingMode') properties in the **scrolling** object to *"virtual"*. These properties specify that UI elements are only rendered when they come into the viewport. Use virtual row rendering when [paging is disabled](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/enabled.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#enabled') or the [pageSize](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/paging/pageSize.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/paging/#pageSize') is more than the default value. Virtual column rendering can be used when there are more than 20 columns outside the viewport.
 
+* **Render elements with complex content last**           
 Cells with complex content (predefined templates, command columns, editors) can be rendered after cells with simple content. To do that, enable the [renderAsync](/api-reference/10%20UI%20Widgets/GridBase/1%20Configuration/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#renderAsync') property. In addition, you can enable [the same property](/api-reference/_hidden/GridBaseColumn/renderAsync.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#renderAsync') for individual columns if they use a content-heavy [cellTemplate](/api-reference/_hidden/dxDataGridColumn/cellTemplate.md '/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate').
```

Comment:
What is predefined templates? I do not know. I think we should give more specific information: "filter row cells, command columns, checkboxes in boolean columns". 
Also, in React users use cellRender and cellComponent instead of cellTemplate. Can we say about this react specific?